test_agent/rust/src/utils.rs

```diff
@@ -0,0 +1,491 @@
+                uattributes.source = Some(wrapper_uri.0).into();
+            }
+        };
+        if let Some(sink_value) = value.get("sink") {
+            if let Ok(wrapper_uri) = serde_json::from_value::<WrapperUUri>(sink_value.clone()) {
+                uattributes.sink = Some(wrapper_uri.0).into();
+            }
+        };
+
+        let Ok(id) = parse_uuid(&value, "id") else {
+            let err_msg = "Error parsing UUID id: ".to_string();
+            return Err(serde::de::Error::custom(err_msg));
+        };
+
+        uattributes.id = MessageField(Some(Box::new(id)));
+
+        if value.get("ttl").and_then(|ttl| ttl.as_str()).is_some() {
+            if let Ok(Some(parsed_ttl)) = parse_u32_field("entity", &value, "ttl") {
+                uattributes.ttl = parsed_ttl.into();
+            } else {
+                error!("Error: Failed to parse ttl as u32");
+            }
+        }
+
+        if value
+            .get("permission_level")
+            .and_then(|permission_level| permission_level.as_str())
+            .is_some()
+        {
+            if let Ok(Some(parsed_permission_level)) =
+                parse_u32_field("entity", &value, "permission_level")
+            {
+                uattributes.permission_level = Some(parsed_permission_level);
+            } else {
+                error!("Error: Failed to parse permission_level as u32");
+            }
+        };
+
+        match parse_string_field("entity", &value, "commstatus") {
+            Ok(commstatus_str) => {
+                if let Some(code) = UCode::from_str(&commstatus_str) {
+                    uattributes.commstatus = Some(code.into());
+                } else {
+                    error!("Failed to parse commstatus string into UCode.");
+                }
+            }
+            Err(err) => {
+                error!("Error parsing commstatus field: {}", err);
+            }
+        }
+
+        debug!(" uattributes.commstatus: {:?}", uattributes.commstatus);
+
+        if let Ok(reqid) = parse_uuid(&value, "reqid") {
+            uattributes.reqid = MessageField(Some(Box::new(reqid)));
+        } else {
+            let err_msg = "Error parsing UUID reqid: ".to_string();
+            error!("{}", err_msg);
+        }
+
+        if let Some(token) = value.get("token") {
+            if let Some(token_str) = token.as_str() {
+                uattributes.token = Some(token_str.to_owned());
+            } else {
+                error!("Error: token is not a string");
+            }
+        };
+        if let Some(traceparent) = value
+            .get("traceparent")
+            .and_then(|traceparent| traceparent.as_str())
+        {
+            uattributes.traceparent = Some(traceparent.to_owned());
+        } else {
+            error!("Error: traceparent is not a string");
+        };
+
+        uattributes.special_fields = SpecialFields::default();
```

Comment:
Oh... so did initializing uattributes with Default not set special_fields...? If so, how annoying...